lxd/storage/drivers/driver_pure_volumes.go

```diff
@@ -186,14 +189,99 @@ func (d *pure) RefreshVolume(vol VolumeCopy, srcVol VolumeCopy, refreshSnapshots
 	return err
 }
 
-// DeleteVolume deletes a volume of the storage device.
-// If any snapshots of the volume remain then this function will return an error.
+// DeleteVolume deletes the volume and all associated snapshots.
 func (d *pure) DeleteVolume(vol Volume, op *operations.Operation) error {
+	volExists, err := d.HasVolume(vol)
+	if err != nil {
+		return err
+	}
+
+	if !volExists {
+		return nil
+	}
+
+	volName, err := d.getVolumeName(vol)
+	if err != nil {
+		return err
+	}
+
+	var host *pureHost
+
+	switch d.config["pure.mode"] {
```

Comment:
can we abstract this whole section for disconnecting a volume into a separate function that deals with the different modes?